File: backend/app/utils/units.py
"""
Unit conversion utilities for imperial and metric measurements.

Supports conversion between:
- Inches (imperial) ↔ Millimeters (metric)
- Inches (imperial) ↔ Centimeters (metric)
"""

from decimal import Decimal
from typing import Union

# Conversion constants
INCHES_TO_MM = Decimal("25.4")
INCHES_TO_CM = Decimal("2.54")
MM_TO_INCHES = Decimal("1") / INCHES_TO_MM
CM_TO_INCHES = Decimal("1") / INCHES_TO_CM


def inches_to_mm(inches: Union[Decimal, float, int]) -> Decimal:
    """
    Convert inches to millimeters.

    Args:
        inches: Measurement in inches

    Returns:
        Measurement in millimeters
    """
    if not isinstance(inches, Decimal):
        inches = Decimal(str(inches))
    return inches * INCHES_TO_MM


def inches_to_cm(inches: Union[Decimal, float, int]) -> Decimal:
    """
    Convert inches to centimeters.

    Args:
        inches: Measurement in inches

    Returns:
        Measurement in centimeters
    """
    if not isinstance(inches, Decimal):
        inches = Decimal(str(inches))
    return inches * INCHES_TO_CM


def mm_to_inches(mm: Union[Decimal, float, int]) -> Decimal:
    """
    Convert millimeters to inches.

    Args:
        mm: Measurement in millimeters

    Returns:
        Measurement in inches
    """
    if not isinstance(mm, Decimal):
        mm = Decimal(str(mm))
    return mm * MM_TO_INCHES


def cm_to_inches(cm: Union[Decimal, float, int]) -> Decimal:
    """
    Convert centimeters to inches.

    Args:
        cm: Measurement in centimeters

    Returns:
        Measurement in inches
    """
    if not isinstance(cm, Decimal):
        cm = Decimal(str(cm))
    return cm * CM_TO_INCHES


def format_measurement(
    value: Union[Decimal, float, int],
    unit_system: str = "imperial",
    precision: int = 2
) -> str:
    """
    Format a measurement with appropriate units.

    Assumes input is always in inches (internal storage format).
    Converts to metric if unit_system is "metric".

    Args:
        value: Measurement value (in inches)
        unit_system: "imperial" or "metric"
        precision: Number of decimal places

    Returns:
        Formatted string with units (e.g., "12.50\"" or "317.5 mm")
    """
    if not isinstance(value, Decimal):
        value = Decimal(str(value))

    if unit_system == "metric":
        mm_value = inches_to_mm(value)
        return f"{mm_value:.{precision}f} mm"
    else:
        return f'{value:.{precision}f}"'


def parse_measurement(
    value_str: str,
    unit_system: str = "imperial"
) -> Decimal:
    """
    Parse a measurement string to internal format (inches).

    Args:
        value_str: Measurement string (e.g., "12.5" or "317.5")
        unit_system: "imperial" or "metric"

    Returns:
        Measurement in inches (internal format)
    """
    # Remove units and whitespace
    clean_str = value_str.strip().replace('"', '').replace('mm', '').replace('cm', '').strip()
    value = Decimal(clean_str)

    # Convert to inches if metric
    if unit_system == "metric":
        # Assume mm if > 100, cm if <= 100 (heuristic)
        if value > 100:
            return mm_to_inches(value)
        else:
            return cm_to_inches(value)
    else:
        return value


def get_unit_label(unit_system: str = "imperial") -> str:
    """
    Get the display label for the unit system.

    Args:
        unit_system: "imperial" or "metric"

    Returns:
        Display label (e.g., "inches" or "mm")
    """
    return "inches" if unit_system == "imperial" else "mm"


def get_unit_symbol(unit_system: str = "imperial") -> str:
    """
    Get the symbol for the unit system.

    Args:
        unit_system: "imperial" or "metric"

    Returns:
        Unit symbol (e.g., "\"" or "mm")
    """
    return '"' if unit_system == "imperial" else "mm"

File: frontend/src/utils/units.ts
/**
 * Unit conversion utilities for imperial and metric measurements.
 *
 * Supports conversion between:
 * - Inches (imperial) ↔ Millimeters (metric)
 * - Inches (imperial) ↔ Centimeters (metric)
 */

export type UnitSystem = 'imperial' | 'metric'

// Conversion constants
const INCHES_TO_MM = 25.4
const INCHES_TO_CM = 2.54
const MM_TO_INCHES = 1 / INCHES_TO_MM
const CM_TO_INCHES = 1 / INCHES_TO_CM

/**
 * Convert inches to millimeters
 */
export function inchesToMm(inches: number): number {
  return inches * INCHES_TO_MM
}

/**
 * Convert inches to centimeters
 */
export function inchesToCm(inches: number): number {
  return inches * INCHES_TO_CM
}

/**
 * Convert millimeters to inches
 */
export function mmToInches(mm: number): number {
  return mm * MM_TO_INCHES
}

/**
 * Convert centimeters to inches
 */
export function cmToInches(cm: number): number {
  return cm * CM_TO_INCHES
}

/**
 * Format a measurement with appropriate units.
 * Assumes input is always in inches (internal storage format).
 * Converts to metric if unitSystem is "metric".
 */
export function formatMeasurement(
  value: number,
  unitSystem: UnitSystem = 'imperial',
  precision: number = 2
): string {
  if (unitSystem === 'metric') {
    const mmValue = inchesToMm(value)
    return `${mmValue.toFixed(precision)} mm`
  } else {
    return `${value.toFixed(precision)}"`
  }
}

/**
 * Parse a measurement string to internal format (inches).
 */
export function parseMeasurement(
  valueStr: string,
  unitSystem: UnitSystem = 'imperial'
): number {
  // Remove units and whitespace
  const cleanStr = valueStr.trim().replace(/['"]/g, '').replace(/mm/g, '').replace(/cm/g, '').trim()
  const value = parseFloat(cleanStr)

  if (isNaN(value)) {
    throw new Error(`Invalid measurement: ${valueStr}`)
  }

  // Convert to inches if metric
  if (unitSystem === 'metric') {
    // Assume mm if > 100, cm if <= 100 (heuristic)
    if (value > 100) {
      return mmToInches(value)
    } else {
      return cmToInches(value)
    }
  } else {
    return value
  }
}

/**
 * Convert a value from internal format (inches) to display format
 */
export function toDisplayValue(
  inches: number | string | null | undefined,
  unitSystem: UnitSystem = 'imperial',
  precision: number = 2
): number {
  // Convert string to number if needed (Decimal from backend serializes as string)
  const numericValue = typeof inches === 'string' ? parseFloat(inches) : inches

  // Handle undefined, null, or NaN values
  if (numericValue === undefined || numericValue === null || isNaN(numericValue)) {
    return 0
  }

  if (unitSystem === 'metric') {
    return parseFloat(inchesToMm(numericValue).toFixed(precision))
  } else {
    return parseFloat(numericValue.toFixed(precision))
  }
}

/**
 * Convert a value from display format to internal format (inches)
 */
export function toInternalValue(
  displayValue: number | string | null | undefined,
  unitSystem: UnitSystem = 'imperial'
): number {
  // Convert string to number if needed
  const numericValue = typeof displayValue === 'string' ? parseFloat(displayValue) : displayValue

  // Handle undefined, null, or NaN values
  if (numericValue === undefined || numericValue === null || isNaN(numericValue)) {
    return 0
  }

  if (unitSystem === 'metric') {
    return mmToInches(numericValue)
  } else {
    return numericValue
  }
}

/**
 * Get the display label for the unit system
 */
export function getUnitLabel(unitSystem: UnitSystem = 'imperial'): string {
  return unitSystem === 'imperial' ? 'inches' : 'mm'
}

/**
 * Get the symbol for the unit system
 */
export function getUnitSymbol(unitSystem: UnitSystem = 'imperial'): string {
  return unitSystem === 'imperial' ? '"' : 'mm'
}

/**
 * Get the step value for input fields based on unit system
 */
export function getInputStep(unitSystem: UnitSystem = 'imperial'): number {
  return unitSystem === 'imperial' ? 0.125 : 1 // 1/8 inch or 1mm
}
